### PR TITLE
TEET-374: As a TEET user I can edit my comments

### PR DESCRIPTION
**Pull request by outamaa**
https://jira.mkm.ee/browse/TEET-374

Woopsie, the image editing is still missing.

### CODE REVIEW

File: app/backend/src/clj/teet/comment/comment_commands.clj [hunk 1]
```diff
@@ -44,29 +44,32 @@
...
+    [:file file-id]
+    (if-let [task-id (ffirst
+                      (d/q '[:find ?task
                              :in $ ?comment
-                             :where [?file :file/comments ?comment]]
+                             :where [?task :task/comments ?comment]]
                            db comment-id))]
-      [:file file-id]
+      [:task task-id]
       nil)))
```
Comment (outamaa) [inline app/backend/src/clj/teet/comment/comment_commands.clj:59]: Replaced `:document` with `:task`.

File: app/backend/src/clj/teet/comment/comment_commands.clj [hunk 2]
```diff
@@ -44,29 +44,32 @@
...
 
+(defn- get-project-id-of-comment [db comment-id]
+  (let [[parent-type parent-id] (comment-parent-entity db comment-id)]
+    (case parent-type
+      :file (project-db/file-project-id db parent-id)
+      :task (project-db/task-project-id db parent-id)
+      (db-api/bad-request! "No such comment"))))
+
```
Comment (outamaa) [inline app/backend/src/clj/teet/comment/comment_commands.clj:67]: Refactored project id fetch into a separate function.

File: app/backend/src/clj/teet/comment/comment_queries.clj [hunk 3]
```diff
@@ -21,6 +22,7 @@
       []
       (->> entity-comments
            attr
+           (meta-query/without-deleted db)
```
Comment (outamaa) [inline app/backend/src/clj/teet/comment/comment_queries.clj:25]: Deleted comments were fetched to the frontend.

File: app/common/src/cljc/teet/authorization/authorization_check.cljc [hunk 4]
```diff
@@ -83,20 +84,33 @@
...
+          (when (and permissions
+                     user
+                     action-permissions)
+            (when (every? (fn [[permission {:keys [link]}]]
+                            (authorized? @app-state/user permission (merge {:project-id project-id
+                                                                            :entity entity}
+
+                                                                           (when link {:link link}))))
+                          (action permissions))
+              component))))]))
```
Comment (outamaa) [inline app/common/src/cljc/teet/authorization/authorization_check.cljc:113]: Changed `when-authorized` into a component that uses React context.

File: app/common/src/cljc/teet/util/collection.cljc [hunk 5]
```diff
@@ -39,6 +39,13 @@
             element))
         collection))
 
+(defn find-by-id
+  "Find first element in `collection` whose `:db/id` is `id`"
+  [id collection]
+  (find-first (comp (partial = id)
+                    :db/id)
+              collection))
+
```
Comment (outamaa) [inline app/common/src/cljc/teet/util/collection.cljc:48]: Didn't end up using this after all, so if this doesn't look useful, I can remove this.

File: app/frontend/src/cljs/teet/comments/comments_controller.cljs [hunk 6]
```diff
@@ -58,9 +72,57 @@
...
            :command          :comment/delete-comment
            :payload          {:comment-id comment-id}
-           :result-event     common-controller/->Refresh})))
+           :success-message (tr [:notifications :comment-deleted])
+           :result-event     (partial ->DeleteCommentResult commented-entity)}))
+
+  DeleteCommentResult
+  (process-event [{:keys [commented-entity]} app]
+    (t/fx app
+          (comments-query commented-entity)))
```
Comment (outamaa) [inline app/frontend/src/cljs/teet/comments/comments_controller.cljs:86]: Reload comments after deleting comment.

File: app/frontend/src/cljs/teet/comments/comments_controller.cljs [hunk 7]
```diff
@@ -58,9 +72,57 @@
...
+            :payload (select-keys edit-comment-data [:db/id :comment/comment])
+            :success-message (tr [:notifications :comment-edited])})))
+
+  SaveEditCommentSuccess
+  (process-event [_ app]
+    (let [commented-entity (-> app :edit-comment-data :comment/commented-entity)]
+      (t/fx (-> app
+                (dissoc :edit-comment-data)
+                (update :stepper dissoc :dialog))
+            (comments-query commented-entity)))))
```
Comment (outamaa) [inline app/frontend/src/cljs/teet/comments/comments_controller.cljs:128]: Breaks my heart to write 40 lines of code to open a textarea dialog box. :)

File: app/frontend/src/cljs/teet/comments/comments_controller.cljs [hunk 8]
```diff
@@ -58,9 +72,57 @@
...
+
+  DeleteCommentResult
+  (process-event [{:keys [commented-entity]} app]
+    (t/fx app
+          (comments-query commented-entity)))
+
+  OpenEditCommentDialog
+  (process-event [{:keys [comment-id commented-entity comment]} app]
+    (-> app
+        (assoc-in [:stepper :dialog] {:type :edit-comment})
```
Comment (Trpal) [inline app/frontend/src/cljs/teet/comments/comments_controller.cljs:90]: Not sure if this stepper should be used as a place for "not stepper" related dialogs? Or it should be renamed, the whole :stepper is from a time when the project navigator was supposed to be a lot different.

Comment (outamaa) [reply app/frontend/src/cljs/teet/comments/comments_controller.cljs:90]: Given how ubiquitous dialogs are, I'd like to have an easy, general mechanism for firing up dialogs. So I'd go with renaming.

**Pull request merged**